# Numpy, lists and tuples practicum 2/3/2022

### Start by importing numpy:

In [1]:
import numpy as np

#### First create an array x from 100 to 1,000 with a step of 2:

In [2]:
x = np.arange( 100, 1000, 2)

In [3]:
print(x)

[100 102 104 106 108 110 112 114 116 118 120 122 124 126 128 130 132 134
 136 138 140 142 144 146 148 150 152 154 156 158 160 162 164 166 168 170
 172 174 176 178 180 182 184 186 188 190 192 194 196 198 200 202 204 206
 208 210 212 214 216 218 220 222 224 226 228 230 232 234 236 238 240 242
 244 246 248 250 252 254 256 258 260 262 264 266 268 270 272 274 276 278
 280 282 284 286 288 290 292 294 296 298 300 302 304 306 308 310 312 314
 316 318 320 322 324 326 328 330 332 334 336 338 340 342 344 346 348 350
 352 354 356 358 360 362 364 366 368 370 372 374 376 378 380 382 384 386
 388 390 392 394 396 398 400 402 404 406 408 410 412 414 416 418 420 422
 424 426 428 430 432 434 436 438 440 442 444 446 448 450 452 454 456 458
 460 462 464 466 468 470 472 474 476 478 480 482 484 486 488 490 492 494
 496 498 500 502 504 506 508 510 512 514 516 518 520 522 524 526 528 530
 532 534 536 538 540 542 544 546 548 550 552 554 556 558 560 562 564 566
 568 570 572 574 576 578 580 582 584 586 588 590 59

### then create an array y from 1,000 to 2,800 with a step of 4:

In [4]:
y = np.arange( 1000, 2800, 4)

In [5]:
print(y)

[1000 1004 1008 1012 1016 1020 1024 1028 1032 1036 1040 1044 1048 1052
 1056 1060 1064 1068 1072 1076 1080 1084 1088 1092 1096 1100 1104 1108
 1112 1116 1120 1124 1128 1132 1136 1140 1144 1148 1152 1156 1160 1164
 1168 1172 1176 1180 1184 1188 1192 1196 1200 1204 1208 1212 1216 1220
 1224 1228 1232 1236 1240 1244 1248 1252 1256 1260 1264 1268 1272 1276
 1280 1284 1288 1292 1296 1300 1304 1308 1312 1316 1320 1324 1328 1332
 1336 1340 1344 1348 1352 1356 1360 1364 1368 1372 1376 1380 1384 1388
 1392 1396 1400 1404 1408 1412 1416 1420 1424 1428 1432 1436 1440 1444
 1448 1452 1456 1460 1464 1468 1472 1476 1480 1484 1488 1492 1496 1500
 1504 1508 1512 1516 1520 1524 1528 1532 1536 1540 1544 1548 1552 1556
 1560 1564 1568 1572 1576 1580 1584 1588 1592 1596 1600 1604 1608 1612
 1616 1620 1624 1628 1632 1636 1640 1644 1648 1652 1656 1660 1664 1668
 1672 1676 1680 1684 1688 1692 1696 1700 1704 1708 1712 1716 1720 1724
 1728 1732 1736 1740 1744 1748 1752 1756 1760 1764 1768 1772 1776 1780
 1784 

### check the shape of x and y:

In [6]:
print( x.shape )
print( y.shape )

(450,)
(450,)


In [7]:
print(len(x))

450


### Create array z equal to the difference of y minus x 

In [8]:
z = np.array(y-x)

### print it:

In [9]:
print(z)

[ 900  902  904  906  908  910  912  914  916  918  920  922  924  926
  928  930  932  934  936  938  940  942  944  946  948  950  952  954
  956  958  960  962  964  966  968  970  972  974  976  978  980  982
  984  986  988  990  992  994  996  998 1000 1002 1004 1006 1008 1010
 1012 1014 1016 1018 1020 1022 1024 1026 1028 1030 1032 1034 1036 1038
 1040 1042 1044 1046 1048 1050 1052 1054 1056 1058 1060 1062 1064 1066
 1068 1070 1072 1074 1076 1078 1080 1082 1084 1086 1088 1090 1092 1094
 1096 1098 1100 1102 1104 1106 1108 1110 1112 1114 1116 1118 1120 1122
 1124 1126 1128 1130 1132 1134 1136 1138 1140 1142 1144 1146 1148 1150
 1152 1154 1156 1158 1160 1162 1164 1166 1168 1170 1172 1174 1176 1178
 1180 1182 1184 1186 1188 1190 1192 1194 1196 1198 1200 1202 1204 1206
 1208 1210 1212 1214 1216 1218 1220 1222 1224 1226 1228 1230 1232 1234
 1236 1238 1240 1242 1244 1246 1248 1250 1252 1254 1256 1258 1260 1262
 1264 1266 1268 1270 1272 1274 1276 1278 1280 1282 1284 1286 1288 1290
 1292 

### multiply x by 2, subtract 100 from y and add all this togeter ( (x*2) + (y-100)); assign to variable test1 

In [10]:
test1 = (x * 2) + (y - 100)

print(test1)

[1100 1108 1116 1124 1132 1140 1148 1156 1164 1172 1180 1188 1196 1204
 1212 1220 1228 1236 1244 1252 1260 1268 1276 1284 1292 1300 1308 1316
 1324 1332 1340 1348 1356 1364 1372 1380 1388 1396 1404 1412 1420 1428
 1436 1444 1452 1460 1468 1476 1484 1492 1500 1508 1516 1524 1532 1540
 1548 1556 1564 1572 1580 1588 1596 1604 1612 1620 1628 1636 1644 1652
 1660 1668 1676 1684 1692 1700 1708 1716 1724 1732 1740 1748 1756 1764
 1772 1780 1788 1796 1804 1812 1820 1828 1836 1844 1852 1860 1868 1876
 1884 1892 1900 1908 1916 1924 1932 1940 1948 1956 1964 1972 1980 1988
 1996 2004 2012 2020 2028 2036 2044 2052 2060 2068 2076 2084 2092 2100
 2108 2116 2124 2132 2140 2148 2156 2164 2172 2180 2188 2196 2204 2212
 2220 2228 2236 2244 2252 2260 2268 2276 2284 2292 2300 2308 2316 2324
 2332 2340 2348 2356 2364 2372 2380 2388 2396 2404 2412 2420 2428 2436
 2444 2452 2460 2468 2476 2484 2492 2500 2508 2516 2524 2532 2540 2548
 2556 2564 2572 2580 2588 2596 2604 2612 2620 2628 2636 2644 2652 2660
 2668 

### calculate the min of test1, its max and its mean:

In [11]:
print(np.min(test1) , np.max(test1) , np.mean(test1))

1100 4692 2896.0


### print the mean of the base 10 logarithm of x

In [12]:
print(np.mean(np.log10(x)))

2.675704070469898


### 1. Create a numpy array x that goes from 1 up to and including 100 (last element must be 100.0) with a step of 0.1. Verify its type using the type command and its size using the .shape command.  Print x. Print its maximum, minimum and mean. Add 5 to x and print it. Print an array whose values are 10 times larger than x.

In [13]:
x = np.arange(1 , 100.1 , 0.1)

In [42]:
list_new = list(range(0,100,10))
print(list_new)

[0, 10, 20, 30, 40, 50, 60, 70, 80, 90]


In [41]:
print(type(x))
print(len(x))
print(x.shape)

<class 'numpy.ndarray'>
991
(991,)


In [15]:
print(x)

[  1.    1.1   1.2   1.3   1.4   1.5   1.6   1.7   1.8   1.9   2.    2.1
   2.2   2.3   2.4   2.5   2.6   2.7   2.8   2.9   3.    3.1   3.2   3.3
   3.4   3.5   3.6   3.7   3.8   3.9   4.    4.1   4.2   4.3   4.4   4.5
   4.6   4.7   4.8   4.9   5.    5.1   5.2   5.3   5.4   5.5   5.6   5.7
   5.8   5.9   6.    6.1   6.2   6.3   6.4   6.5   6.6   6.7   6.8   6.9
   7.    7.1   7.2   7.3   7.4   7.5   7.6   7.7   7.8   7.9   8.    8.1
   8.2   8.3   8.4   8.5   8.6   8.7   8.8   8.9   9.    9.1   9.2   9.3
   9.4   9.5   9.6   9.7   9.8   9.9  10.   10.1  10.2  10.3  10.4  10.5
  10.6  10.7  10.8  10.9  11.   11.1  11.2  11.3  11.4  11.5  11.6  11.7
  11.8  11.9  12.   12.1  12.2  12.3  12.4  12.5  12.6  12.7  12.8  12.9
  13.   13.1  13.2  13.3  13.4  13.5  13.6  13.7  13.8  13.9  14.   14.1
  14.2  14.3  14.4  14.5  14.6  14.7  14.8  14.9  15.   15.1  15.2  15.3
  15.4  15.5  15.6  15.7  15.8  15.9  16.   16.1  16.2  16.3  16.4  16.5
  16.6  16.7  16.8  16.9  17.   17.1  17.2  17.3  1

In [16]:
print( np.min(x) , np.max(x) , np.mean(x))

1.0 100.00000000000009 50.500000000000036


In [17]:
x = x + 5
print(x[0])

6.0


In [18]:
print( x * 10)

[  60.   61.   62.   63.   64.   65.   66.   67.   68.   69.   70.   71.
   72.   73.   74.   75.   76.   77.   78.   79.   80.   81.   82.   83.
   84.   85.   86.   87.   88.   89.   90.   91.   92.   93.   94.   95.
   96.   97.   98.   99.  100.  101.  102.  103.  104.  105.  106.  107.
  108.  109.  110.  111.  112.  113.  114.  115.  116.  117.  118.  119.
  120.  121.  122.  123.  124.  125.  126.  127.  128.  129.  130.  131.
  132.  133.  134.  135.  136.  137.  138.  139.  140.  141.  142.  143.
  144.  145.  146.  147.  148.  149.  150.  151.  152.  153.  154.  155.
  156.  157.  158.  159.  160.  161.  162.  163.  164.  165.  166.  167.
  168.  169.  170.  171.  172.  173.  174.  175.  176.  177.  178.  179.
  180.  181.  182.  183.  184.  185.  186.  187.  188.  189.  190.  191.
  192.  193.  194.  195.  196.  197.  198.  199.  200.  201.  202.  203.
  204.  205.  206.  207.  208.  209.  210.  211.  212.  213.  214.  215.
  216.  217.  218.  219.  220.  221.  222.  223.  2

### 2. Create an array y that is the base 10 logarithm of x. Print y. Print the 10th to 24th elements of x and y. Create an array z that is the same as y. Change the 30th to 140th elements of z to be the 1/10 of their current value and the 500th to 600th elements of z to be 10 times larger than their current value. Print z from the 600th element to the 100th element going backwards with a step of 50 (so the 600th, 550th, 500th etc element).   

### 3. Create array base1 that goes from 100 to 600 with a step of 2. Create array base2 that goes from 75.5 to 0.5 with a step of -0.3. Assuming that base1 is the one side of a (orthogonal parallelogram) field and base2 the other side, calculate the surface area of all fields (field_surface). How many fields are there in your sample? How many unique surface areas are there in your sample?

In [19]:
base1 = np.arange( 100 , 600 , 2 )
base2 = np.arange( 75.5 , 0.5 , -0.3 )

In [20]:
print( len( base1 ) , len( base2 ))

250 250


In [21]:
field_surface = base1 * base2 


print(field_surface)

print(len(field_surface))

print(np.unique(field_surface))

print( len(np.unique(field_surface)))

[ 7550.   7670.4  7789.6  7907.6  8024.4  8140.   8254.4  8367.6  8479.6
  8590.4  8700.   8808.4  8915.6  9021.6  9126.4  9230.   9332.4  9433.6
  9533.6  9632.4  9730.   9826.4  9921.6 10015.6 10108.4 10200.  10290.4
 10379.6 10467.6 10554.4 10640.  10724.4 10807.6 10889.6 10970.4 11050.
 11128.4 11205.6 11281.6 11356.4 11430.  11502.4 11573.6 11643.6 11712.4
 11780.  11846.4 11911.6 11975.6 12038.4 12100.  12160.4 12219.6 12277.6
 12334.4 12390.  12444.4 12497.6 12549.6 12600.4 12650.  12698.4 12745.6
 12791.6 12836.4 12880.  12922.4 12963.6 13003.6 13042.4 13080.  13116.4
 13151.6 13185.6 13218.4 13250.  13280.4 13309.6 13337.6 13364.4 13390.
 13414.4 13437.6 13459.6 13480.4 13500.  13518.4 13535.6 13551.6 13566.4
 13580.  13592.4 13603.6 13613.6 13622.4 13630.  13636.4 13641.6 13645.6
 13648.4 13650.  13650.4 13649.6 13647.6 13644.4 13640.  13634.4 13627.6
 13619.6 13610.4 13600.  13588.4 13575.6 13561.6 13546.4 13530.  13512.4
 13493.6 13473.6 13452.4 13430.  13406.4 13381.6 1335

### One of the great features of numpy is that it also allows you to read data in an array in a single command. You do that with np.loadtxt( "name_of_file" ) [see also: https://numpy.org/doc/stable/reference/generated/numpy.loadtxt.html ] or np.genfromtxt("name_of_file") [see also: https://numpy.org/doc/stable/reference/generated/numpy.genfromtxt.html ]. E.g., get file *my_first_data.dat* from the Webcouses/demos/ . Open it with a text editor to see what it looks like. Read it in an array *all_input_data* with loadtxt or genfromtxt. Check what your array looks like. Print all times each car traveled. Print all distances the cars traveled. Calculate the distance each car traveled **in a single line** and asign it to array *distances_traveled*. Print *distances_traveled*

In [22]:
all_input_data = np.loadtxt( 'my_first_data.dat', comments= '#' )

#all_input_data2 = np.genfromtxt

In [23]:
# Car_number  Time [hr]  Speed [mph]
print( all_input_data )

[[  1.    5.2 120.2]
 [  2.    3.1 110. ]
 [  3.    2.4 100.5]
 [  4.    5.1  65.3]
 [  5.    2.4 119.6]]


In [24]:
print( all_input_data[ : , 2]) #prints only the velocity

[120.2 110.  100.5  65.3 119.6]


In [25]:
distances_traveled = all_input_data[ : , 2] * all_input_data[ : , 1]

print(distances_traveled)

[625.04 341.   241.2  333.03 287.04]


### Now let's try the genfromtxt with data: *test_genfromtxt.dat*. Read data into variable student_data. Print the IDs of the students. Print the average quiz  grade for student 202204. Print the average exam quiz and the IDs of all students that passed the class with a C- or better ( average total of more than 7.0).

In [26]:
stu_data = np.genfromtxt('test_genfromtxt.dat')

In [27]:
print(stu_data)

[[2.02201e+05 8.10000e+00 8.40000e+00 8.20000e+00]
 [2.02202e+05 7.50000e+00 7.00000e+00 7.30000e+00]
 [2.02203e+05 7.90000e+00 8.50000e+00 8.40000e+00]
 [2.02204e+05 9.50000e+00 9.70000e+00 9.70000e+00]
 [2.02205e+05 5.30000e+00 6.00000e+00 5.90000e+00]
 [2.02206e+05 5.10000e+00 7.10000e+00 6.80000e+00]]


In [28]:
#student ID AVG-Grade-Quiz AVG-Grade_Exam AVG-Total


In [29]:
print(stu_data[ : , 0])

[202201. 202202. 202203. 202204. 202205. 202206.]


In [30]:
print( stu_data[ 3 , 1 ] ) 

9.5


In [31]:
print(np.where(stu_data[ : , 3 ]>7 ) ) 
print( stu_data[ : , 3 ]>7)

(array([0, 1, 2, 3], dtype=int64),)
[ True  True  True  True False False]


In [50]:
print( stu_data[stu_data[ : , 3 ]>7,1])

[8.1 7.5 7.9 9.5]


### 4. You just opened a new savings account that earns N percent interest per year. The interest that you earn is paid at the end of the year, and is added to the balance of the savings account. Write a *function end_of_year_interest* that takes as input the interest N and a single list *deposits_list* that containts the amount of money deposited into the account from the user at the start of years 1, 2 and 3 (assume this is the only deposit to the account). The function should then calculate the amount in the savings account at the end of years 1, 2 and 3 and print it out. Call the function for an interest rate of 2% and deposits of 100, 300 and 500. 


In [33]:
#put your function here:

def end_of_year_interest( N , deposits_list):
    """Calculates the amount of savings in account at the end of the years 1,2, and 3.
    INPUT: rate[%] , list of deposits
    OUTPUT: list of deposits (and prints of deposits)"""
    
    year1_interest = deposits_list[ 0 ] * ( N / 100 )
    year1_final = deposits_list[ 0 ] * (1 + ( N / 100 ) )
    
    year2_interest = ( year1_final + deposits_list[ 1 ] ) * ( N / 100 )
    year2_final = ( year1_final + deposits_list[ 1 ] ) * ( 1 + ( N / 100 ) )
    
    year3_interest = ( year2_final + deposits_list[ 2 ] ) * ( N / 100 ) 
    year3_final = ( year2_final + deposits_list[ 2 ] ) * ( 1 + ( N / 100 ) )
    
    print(f'At the end of each year have { year1_final , year2_final , year3_final }')
    
    return [ year1_final , year2_final , year3_final ]

In [34]:
#call it here:

interest = 2

deposits = [ 100 , 300 , 500 ]

final_3years = end_of_year_interest( interest , deposits )

print(final_3years)



At the end of each year have (102.0, 410.04, 928.2408)
[102.0, 410.04, 928.2408]


In [35]:
print(N)

NameError: name 'N' is not defined

### 5. Create a program that calculates how fast an object moves [ m/s ] when falling to the ground from a height h. Assume that the body free falls with initial speed 0 m/s, so that u$_{final}$ = $\sqrt(2 * g * h)$, where g = 9.8 $m/s^2$ and h is in m. Start by creating a numpy array of heights from 3 m to 100m with a step of 2.5 (average height of a floor). Create a variable u_final that calculates the final speed of the falling object from **all** heights in **one** line. Print it

In [ ]:
#make the array of the heights

#define g

# the expression you want is u_final = np.sqrt( 2 * g * heights) so:

h = np.arange( 3 , 102 , 2.5 )

g = 9.8 # m/s^2

u_final = np.sqrt(2 * g * h) 

In [ ]:
print(u_final)

### 6. Create a list *test_list_1* that contains the numbers 1, 2, 5, 7, 22, 34, 42. Print it. Change 22 to -10 (don't just copy-paste and change it in the series, access the appropriate element of test_list_1 and change it). Print the mutated list. Add elements 96 and 102 to the end of the list. Add 0 to the start. 

In [ ]:
list_1 = [ 1 , 2 , 5 , 7 , 22 , 34 , 42 ]
print(list_1)

In [ ]:
list_1[4] = -10

In [ ]:
print(list_1)

In [ ]:
list_1.extend( [ 96 , 102 ])

print(list_1)

In [ ]:
list_1 = [0] + list_1
print(list_1)

### 7. Create a tuple *trip_time* that contains the amount of hours a trip takes. Create a tuple *trip_distance* that contains the distance of a given trip in km. <br> Populate *trip_time* with numbers: 1.5, 3.2, 5.6, 14.3, 26.4 and *trip_distance* with 180, 320, 728, 1573, 3326. Calculate the average speed (remember it is distance/time) of each trip. Can you use plain division of the tuples in a single line to do it? Why/why not? <br> After each calculation store the average speed in a tuple called *trip_average_speed* ("append" each calculation to the previous tuple). Is the tuple you have at the end, the same as at the start? Why/why not? <br> You just realized you made a typo, and the 728 should have been 718! Change the tuple accordingly (and properly) and recalculate and print the new *trip_average_speed* tuple. 

In [ ]:
#make tuples for time and distance:


In [ ]:
#try:


In [ ]:
#make tuple for speed:


In [ ]:
# print final speed tuple
print( trip_average_speed)  # no, the final trip_average_speed is not the same as the start, 
                            #you can check with ID
                            # remember that tuples are immutable, every change means we create a new tuple

In [ ]:
# print final speed tuple


In [ ]:
### also, yes, you can kind of do it in a single line but not in a way we have seen yet, e.g.:
#from operator import truediv 
#trip_average_test_one_line = tuple( map( truediv, trip_distance , trip_time ) )  
#print( trip_average_test_one_line )

In [ ]:
# now let's change the element; remember tuples are immutable!
#inital was trip_distance = ( 180, 320, 728, 1573, 3326 ) --> need to change trip_distance[ 2 ]


In [ ]:
print( trip_distance )

In [ ]:
# you can now just do the same trick, looks a bit ugly, for the average speed:
trip_average_speed = trip_average_speed[ 0 : 2] + \
                     ( trip_distance[ 2 ] / trip_time [ 2 ], ) + \
                     trip_average_speed[ 3: ]

In [ ]:
print( trip_average_speed )

### 8. It's storm season, so let's create a code that calculates the distance to where lightning strucks. Create function *thunderstruck* that calculates the distance from you of the point where lightning struck. Remember that lightning travels with the speed of light, so you can assume you see it instantaneously (as soon as it happens). Thunder on the other hand travels with the speed of sound, at $u_{sound} = 343$  m/s or $1088$ feet/s.  Remember to write an informative docstring. 

1) Discuss: what should the input of the function be? What should be its output?

2) Print the docstring

3) You hear the roar of thunder 5s after you see the lightning. How far away was the strike (in feet)?

4) You hear the roar of thunder 12s after you saw the lightning. Where did it hit in miles ( 1miles = 5280 feet )?


### 9. Open a text editor and save it as *all_my_functions.py*. Copy in there the function *acceleration* from HW your HW and from this practicum the functions *end_of_year_interest* and *thunderstruck*. Change the names of the functions by adding a *_fromfile* at the end of each. <br> In a single command import *acceleration_fromfile* and *enf_of_year_interest_fromfile* and *thunderstruck_fromfile*. Call the latter for the case you hear a thunder 10s after you see the lightning. 

#### Note that if you are using numpy in these functions you will need to start your all_my_functions.py with importing numpy as np.

### C. The Body Mass Index is one of the available measures of how healthy a person is. It is defined as the ratio of a person's weight in kg, over the square of their height in m. Make a function BMI that uses the height and weight of a person and returns their BMI. 

### Make a list heights that contains the heights of people, and a list weights that containts their weights. Populate them with heights: 1.60, 2.10 and 1.80 and weights 60, 120 and 81.

### Call the function BMI for all three people and print their BMI.

###  Create a list *student_h* that contains the heights of students in feet and inches. A person being 5' 6" would be noted as 5.06, a person being 5' 10" would be noted as 5.10. Populate it with the following values: 5.01, 4.10, 6.02, 5.07, 6.11.  Create a function ft_to_m_cm that takes as input a value from *student_h* and returns each height in a list [m, cm] (round cm up). Call the function for all values of *student_h* and print the results with an informative sentence (e.g., Student 1 has a height of 1 m 10 cm ) <br>  (1 ft ~= 0.3048 m and 1 inch ~= 0.0254 m and 1m has 100cm). 

In [ ]:
#make student_h list:

In [ ]:
#make your function:


In [ ]:
#calculate the heights in m and cm and print with informative statement:

### Now change the function (make a copy function ft_to_m2) to accept the height as a list [height_ft, height_inches] and adapt the function accordingly...Call it for a student with a height 6' 11" and a student with a height of 5' 6"

In [ ]:
#make function that takes as input a LIST with feet, inches

In [ ]:
#call it for student 6'11":

In [ ]:
#call it for student 5' 6":